From: Tim Lunn <tim@feathertop.org>
Date: Sat, 22 Mar 2014 10:23:12 +1100
+Subject: Add hooks to run prime scripts (as root) before and after a
+ gdm-session
 
+This allows enabling and disabling the dGPU on log out.
---
 common/gdm-common.c    |    2 +-
 daemon/gdm-session.c   |    6 ++++++
 daemon/gdm-x-session.c |    5 +++++
 data/Makefile.am       |   24 ++++++++++++++++++++++++
 data/Prime             |    6 ++++++
 data/PrimeOff          |    6 ++++++
 6 files changed, 48 insertions(+), 1 deletion(-)
 create mode 100644 data/Prime
 create mode 100644 data/PrimeOff

Index: gdm3-3.28.2/common/gdm-common.c
===================================================================
--- gdm3-3.28.2.orig/common/gdm-common.c
+++ gdm3-3.28.2/common/gdm-common.c
@@ -706,7 +706,7 @@
                 goto out;
         }
 
-        env = gdm_get_script_environment (username,
+        env = gdm_get_script_environment ((g_strcmp0 (username, "root") == 0)?NULL:username,
                                           display_name,
                                           display_hostname,
                                           display_x11_authority_file);
Index: gdm3-3.28.2/daemon/gdm-session.c
===================================================================
--- gdm3-3.28.2.orig/daemon/gdm-session.c
+++ gdm3-3.28.2/daemon/gdm-session.c
@@ -2865,6 +2865,12 @@
 
         g_list_free_full (self->priv->outside_connections, g_object_unref);
         self->priv->outside_connections = NULL;
+
+        /* Run PrimeOff after the session is closed */
+        gdm_run_script (GDMCONFDIR "/PrimeOff", "root",
+                        NULL,
+                        NULL,
+                        NULL);
 }
 
 void
Index: gdm3-3.28.2/daemon/gdm-x-session.c
===================================================================
--- gdm3-3.28.2.orig/daemon/gdm-x-session.c
+++ gdm3-3.28.2/daemon/gdm-x-session.c
@@ -614,6 +614,11 @@
 
         g_debug ("Running X session");
 
+        gdm_run_script (GDMCONFDIR "/Prime", "root",
+                        state->display_name,
+                        NULL, /* hostname */
+                        state->auth_file);
+
         launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_NONE);
 
         if (state->environment != NULL) {
Index: gdm3-3.28.2/data/Makefile.am
===================================================================
--- gdm3-3.28.2.orig/data/Makefile.am
+++ gdm3-3.28.2/data/Makefile.am
@@ -10,6 +10,8 @@
 initdir = $(gdmconfdir)/Init
 postdir = $(gdmconfdir)/PostSession
 predir = $(gdmconfdir)/PreSession
+primedir = $(gdmconfdir)/Prime
+primeoffdir = $(gdmconfdir)/PrimeOff
 postlogindir = $(gdmconfdir)/PostLogin
 workingdir = $(GDM_WORKING_DIR)
 xauthdir = $(GDM_XAUTH_DIR)
@@ -173,6 +175,8 @@
 	Xsession.in 		\
 	Init.in 		\
 	PreSession.in 		\
+	Prime			\
+	PrimeOff      		\
 	PostSession.in 		\
 	PostLogin 		\
 	org.gnome.login-screen.gschema.xml.in \
@@ -237,6 +241,8 @@
 	$(DESTDIR)$(initdir)/Default \
 	$(DESTDIR)$(postlogindir)/Default.sample \
 	$(DESTDIR)$(predir)/Default \
+	$(DESTDIR)$(primedir)/Default \
+	$(DESTDIR)$(primeoffdir)/Default \
 	$(DESTDIR)$(postdir)/Default \
 	$(DESTDIR)$(sysconfdir)/dconf/db/gdm \
 	$(DESTDIR)$(sysconfdir)/dconf/profile/gdm \
@@ -283,6 +289,24 @@
 	fi
 	$(INSTALL_SCRIPT) PreSession $(DESTDIR)$(predir)/Default
 
+	if test '!' -d $(DESTDIR)$(primedir); then \
+		$(mkinstalldirs) $(DESTDIR)$(primedir); \
+		chmod 755 $(DESTDIR)$(primedir); \
+	fi
+	-if test -f $(DESTDIR)$(primedir)/Default; then \
+		cp -f $(DESTDIR)$(primedir)/Default $(DESTDIR)$(primedir)/Default.orig; \
+	fi
+	$(INSTALL_SCRIPT) Prime $(DESTDIR)$(primedir)/Default
+
+	if test '!' -d $(DESTDIR)$(primeoffdir); then \
+		$(mkinstalldirs) $(DESTDIR)$(primeoffdir); \
+		chmod 755 $(DESTDIR)$(primeoffdir); \
+	fi
+	-if test -f $(DESTDIR)$(primeoffdir)/Default; then \
+		cp -f $(DESTDIR)$(primeoffdir)/Default $(DESTDIR)$(primeoffdir)/Default.orig; \
+	fi
+	$(INSTALL_SCRIPT) PrimeOff $(DESTDIR)$(primeoffdir)/Default
+
 	if test '!' -d $(DESTDIR)$(postdir); then \
 		$(mkinstalldirs) $(DESTDIR)$(postdir); \
 		chmod 755 $(DESTDIR)$(postdir); \
Index: gdm3-3.28.2/data/Prime
===================================================================
--- /dev/null
+++ gdm3-3.28.2/data/Prime
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+primeoffload=/sbin/prime-offload
+if [ -f "$primeoffload" ]; then
+    $primeoffload
+fi
Index: gdm3-3.28.2/data/PrimeOff
===================================================================
--- /dev/null
+++ gdm3-3.28.2/data/PrimeOff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+primeswitch=/sbin/prime-switch
+if [ -f "$primeswitch" ]; then
+    $primeswitch
+fi
